Fixes #28584 - refactor API middleware payload

Ron-Lavi opened the pull request:
currently the returned payload is a mix of data and payload, looks like: 
`payload: { ...payload, ...data }`
which is quite risky and may not be the best approach to handle API responses,

in this refactor I will split it into `response` which will store the API's success or error responses, and `payload` which will pass the same payload that was passed to the API action.
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

mmoll commented:
[test katello]

tbrisker reviewed (commented):
This is quite confusing imho, and doesn't solve the issue of mixing the request and response data. 
Why not change the actions to something like this:
```js
{
type: actionTypes.whatever.
request: {url:..., headers: ..., params: ...},
response: { error:..., ...data}
}
```
that way it is similar to the objects we would see in the controller (also named `request` and `response`), it is clear what came from which side, and we don't have to dig in to the object like `payload.results.results.data` etc. but rather just `response.data`

Ron-Lavi commented:
@tbrisker the API action stays as it was and what changes is how the reducer gets the data,
imho it solves the mixing of the two:
the `params` key is storing the `request.params` ,
and the `results` stands for the `response.data`

Usually we don't need to deal with things like `url` or `headers` in our store,
I don't mind to rename it to `response` and `requestParams`, if it makes more sense.

tbrisker commented:
> Usually we don't need to deal with things like url or headers in our store,

Would we ever, even if just for debugging purpose? if not i'm fine with just `requestParams`, otherwise it would be better imho to keep the whole request with all it's attributes and access `request.params`.

> I don't mind to rename it to `response` and `requestParams`, if it makes more sense.

My main point was getting rid of the mixed `payload` object and breaking it to two separate objects, each with a clear meaning. 

Also, i was trying to avoid the deep nesting such as `results.results.data` etc as is currently the case in several of the files you changed. this should be either flat under the `response` object or at most `response.data`.

Ron-Lavi commented:
> Would we ever, even if just for debugging purpose?

For debugging purpose we could check on the API middleware where the API call is taking place,
there you could inspect all of the headers and params that the API request receives, same for the response.

> Also, i was trying to avoid the deep nesting such as results.results.data etc as is currently the case in several of the files you changed. this should be either flat under the response object or at most response.data.

Thanks for explaining, because I stored `response.data` as `results`,
what happened in some of the reducers was: 
`state.setIn([params.controller, 'results'], results.results)`

where `results.results` actually means `response.data.results`
I guess that if we will rename it to `response` it will look better in our code: 
`response.results` in this case.

Ron-Lavi commented:
spoke about it offline with @sharvit,
and it might be a good thing to keep the `payload` that was sent through the API action as `payload`,
because this is something we usually expect from a `payload` to be.
and store the API response `data` as `response` key, e.g:

```js
// action
dispatch({
  type: API_OPERATIONS.GET,
  url: 'some-url',
  payload: {
    controller: 'some-controller'
  }
})
``` 

```js
// reducer
  export default (state = initialState, action) => {
  const { payload, response } = action;

  switch (action.type) {
    case BOOKMARKS_REQUEST:
      return state.setIn([payload.controller, 'results'], response.results)
```

Ron-Lavi commented:
updated with the new convention, having `payload` and `response` separated.
the `error` imho should live inside the `response` too.

tbrisker commented on webpack/assets/javascripts/react_app/components/Bookmarks/__tests__/__snapshots__/integration.test.js.snap:
```diff
@@ -35,6 +35,9 @@ Object {
           "controller": "hosts",
+        },
+        "response": Object {
+          "controller": "hosts",
```
this doesn't seem right?

tbrisker commented on webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/integration.test.js:
```diff
@@ -28,7 +28,11 @@ describe('Bookmark form integration test', () => {
       type: BOOKMARKS_SUCCESS,
-      payload: { ...response.data, item },
+      payload: {
+        controller: response.data.controller,
```
this seems odd, why are we using the controller from the response data and not from the props?

tbrisker commented on webpack/assets/javascripts/react_app/components/common/table/reducers/createTableReducer.js:
```diff
@@ -12,6 +12,7 @@ const initState = Immutable({
 
 const createTableReducer = tableID => (state = initState, action) => {
   const { REQUEST, FAILURE, SUCCESS } = createTableActionTypes(tableID);
+  const { payload, response } = action;
```
just a thought - since we're destructuring the action, should we also do that for the `type`? maybe in the method signature even? (also for the rest of the reducers)

tbrisker commented on webpack/assets/javascripts/react_app/redux/reducers/hosts/storage/vmware.js:
```diff
@@ -123,7 +123,7 @@ export default (state = initialState, { type, payload }) => {
+        .set('datastores', response.results)
         .set('datastoresLoading', false);
     case STORAGE_VMWARE_DATASTORES_FAILURE:
       return state.set('datastoresError', payload.error.message);
```
should this be ?
```suggestion
       return state.set('datastoresError', reponse.error.message);
```
I think there are some other points here where you need to switch the two

tbrisker commented on webpack/assets/javascripts/react_app/redux/reducers/statistics/__snapshots__/statistics.test.js.snap:
```diff
@@ -27,12 +27,7 @@ exports[`statistics reducer should handle STATISTICS_DATA_SUCCESS 1`] = `
-          2,
-        ],
-      ],
+      "data": undefined,
```
This seems wrong

sharvit replied:
Doesn't feel right

tbrisker commented on webpack/assets/javascripts/react_app/redux/reducers/statistics/index.js:
```diff
@@ -10,20 +10,20 @@ const initialState = Immutable({
     case STATISTICS_DATA_REQUEST:
       return state.setIn(['charts', payload.id], payload);
     case STATISTICS_DATA_SUCCESS:
       return state.setIn(['charts', payload.id], {
```
should we be using the id from the payload or from the response (here and elsewhere)?

Ron-Lavi replied:
imo we should use it from the payload, not sure if every response returns the id too.

tbrisker reviewed (changes requested):
This looks much better, see inline comments. 
I still think `payload` isn't as clear as to where it came from. It could refer to the payload sent to the server, the payload returned from the server, or the payload passed around between different components/reducers/actions/etc. If we use `request` or `requestParams`, it is clear that this is what the client asked the server for.

sharvit commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -13,12 +13,14 @@ export const get = async (payload, url, store, actionTypes) => {
       type: actionTypes.FAILURE,
-      payload: { ...payload, error },
+      payload,
+      response: { error },
```
Would it make sense changing it to `response: error`?
For the same reason you don't do `response: { data }`

So if the type is `SUCCESS` you will receive the data as the `response` and if the type is `FAILURE` you will receive the error as the `response`.

Ron-Lavi replied:
done

sharvit reviewed (changes requested):
Thanks @laviro

Ron-Lavi commented:
Thanks, updated with your suggestions @tbrisker @sharvit,

> If we use request or requestParams, it is clear that this is what the client asked the server for.

I agree, but i got mixed feelings about it
because I feel that in the redux flow it's expected that the `payload` that was passed in the `action`, will be accessible also in the `reducer`, which is the same flow we see in other actions.
any other thoughts @sharvit ?

sharvit commented:
> because I feel that in the redux flow it's expected that the `payload` that was passed in the `action`, will be accessible also in the `reducer`, which is the same flow we see in other actions.

I tend to agree. The interface the consumer uses is redux-actions and not `axios` or `fetch`. From a consumer point of view, the `payload` is the request and I don't really care if it's sent as a `query-params` or as the `request.body`, the middleware can decide based on the http method.

sharvit commented:
@laviro can you please check the error in jenkins?

ronlavi2412 committed:
Fixes #28584 - refactor API middleware payload

split the response and payload which are passed
from the middleware to the reducers.

Ron-Lavi commented:
Thanks @sharvit, models page got broken and got fixed now,
I checked also:
- `FACT_CHART`
- `BOOKKMARKS`
- `STATISTICS`
- `TABLES`
- `POWER STATUS`

and they seem to work

sharvit reviewed (approved):
Thanks @laviro, LGTM :+1:

tbrisker reviewed (approved):
Thanks @laviro !
